### PR TITLE
- Removed debug assertion from Rdb_sst_file_ordered::put when keys ar…

### PR DESCRIPTION
**Pull request by george-lorch**
…e out of order as now there is proper SQL error handling for the situation.

- Extended rocksdb.bulk_load_errors.test to induce situation that would lead to debug assertion in  Rdb_sst_file_ordered::put and recorded result.
- Fixed assertion in rocksdb_update_bulk_load by converting it to a 'check' function which can take action and return a failure.
- Fixed up rocksdb.bulk_load_errors.test to remove check for assertion on overlapping keys and just check for proper error return and validate residual state of rocksdb_bulk_load and table.
- Fixed code paths that would call my_print_error to a client where no client exists, this would cause an assertion in the server code.
- Altered rocksdb.bulk_load_errors.test to watch for new errors in log and re-recorded.
- Fixed edge case where if a handler is closed while a bulk load operation with a fault is in process ~ha_rocksdb would assert. Added another sql_print_error and test case to rocksdb.bulk_load_errors.test.

### CODE REVIEW

**Comment by george-lorch** [comment]
Hey team, this issue https://github.com/facebook/mysql-5.6/issues/680 is I think mostly addressed in this PR. It has been a big drag on our QA team and will be a pretty big issue for users to have the server asserting rather than reporting errors.

I am pretty sure some of the changes are not really 'friendly' to the overall MyRocks design, but wanted to at least get something sent in for review, discussion, and eventual clean up. Please provide feedback on anything I may have missed or alternate approaches to solving the issues. Thanks!

**Comment by george-lorch** [comment]
Something that I was considering before but didn't carry through in the initial patch is to make sure that any bulk load related sql_print_error contains the database and table names and possibly even the query that triggered the error.

File: mysql-test/suite/rocksdb/t/bulk_load_errors.test [hunk 1]
```diff
@@ -21,21 +27,32 @@ INSERT INTO t1 VALUES(2);
...
 
---exec grep "RocksDB: Error 197 finalizing last SST file while setting bulk loading variable" $MYSQLTEST_VARDIR/log/mysqld.1.err | cut -d] -f2
---exec echo "" >$MYSQLTEST_VARDIR/log/mysqld.1.err
+SHOW VARIABLES LIKE 'rocksdb_bulk_load';
+
+SELECT * FROM t1;
+
+--let SEARCH_PATTERN=RocksDB: Error 198 finalizing last SST file while setting bulk loading variable
+--source include/search_pattern_in_file.inc
+--exec echo " " > $SEARCH_FILE
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/bulk_load_errors.test]: Restart again with a different SEARCH_FILE instead?

File: mysql-test/suite/rocksdb/t/bulk_load_errors.test [hunk 2]
```diff
@@ -21,21 +27,32 @@ INSERT INTO t1 VALUES(2);
...
-# restart the crashed server
---exec echo "restart" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+SELECT * FROM t1;
 
-# Make sure the error exists in the .err log and then restart the server
---enable_reconnect
---source include/wait_until_connected_again.inc
+--let SEARCH_PATTERN=RocksDB: Error 198 finalizing last SST file while disconnecting
+--source include/search_pattern_in_file.inc
+--exec echo " " > $SEARCH_FILE
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/bulk_load_errors.test]: Likewise?

File: mysql-test/suite/rocksdb/t/bulk_load_errors.test (hunk 1 above)
Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/bulk_load_errors.test]: Slow the test down with no gain?

Comment (laurynas-biveinis) [reply mysql-test/suite/rocksdb/t/bulk_load_errors.test]: In order not to limit portability of the testcase: IIRC, on Windows the "exec echo" would fail to truncate the file because mysqld opens it without allowing shared I/O for any subsequent process trying to open it again. But, this change would be more of a favor for MariaDB MyRocks, as they are the ones with Windows build, and maybe they are in a better position to suggest a patch here.

Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/bulk_load_errors.test]: Hmm, true, that is a fair point.

Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/bulk_load_errors.test]: OK, changed to restarting the server with fresh logs for each search, test takes ~75x longer to run (was ~80ms locally before but is now ~6100ms) but should be safe for any 'doze users.

Comment (laurynas-biveinis) [reply mysql-test/suite/rocksdb/t/bulk_load_errors.test]: The first and the final restarts are required for error log handling correctness on all platforms, thus I hope that Windows adds only a ~2x slowdown :)

File: storage/rocksdb/ha_rocksdb.cc [hunk 3]
```diff
@@ -11973,23 +11973,30 @@ void rocksdb_set_collation_exception_list(THD *const thd,
...
-void rocksdb_set_bulk_load(THD *const thd, struct st_mysql_sys_var *const var
-                                               MY_ATTRIBUTE((__unused__)),
-                           void *const var_ptr, const void *const save) {
-  Rdb_transaction *&tx = get_tx_from_thd(thd);
+int rocksdb_check_bulk_load(THD *const thd,
+                            struct st_mysql_sys_var *var
+                                MY_ATTRIBUTE((__unused__)),
+                            void *save,
+                            struct st_mysql_value *value) {
+    Rdb_transaction *&tx = get_tx_from_thd(thd);
```
Comment (hermanlee) [inline storage/rocksdb/ha_rocksdb.cc]: Nit: These look like whitespace changes. It should be two spaces per indent.

File: storage/rocksdb/ha_rocksdb.cc [hunk 4]
```diff
@@ -11973,23 +11973,30 @@ void rocksdb_set_collation_exception_list(THD *const thd,
...
+                rc);
+            THDVAR(thd, bulk_load) = 0;
+            return 1;
+        }
     }
-  }
 
-  *static_cast<bool *>(var_ptr) = *static_cast<const bool *>(save);
+    long long intbuf;
+    value->val_int(value, &intbuf);
```
Comment (hermanlee) [inline storage/rocksdb/ha_rocksdb.cc]: The test rocksdb_sys_vars.bulk_load_basic is failing to use "on" as 1. I think since it's a bool, value could be a string and needs to evaluate "ON", "OFF" to 1 and 0 too.

**Review by hermanlee — CHANGES_REQUESTED** [review]
Looks great overall. Just need a fix for the rocksdb_sys_vars.rocksdb_bulk_load_basic unit test.

File: storage/rocksdb/ha_rocksdb.cc (hunk 3 above)
Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc]: Oops, sorry, fat-finger on my part. Will be fixed.

File: storage/rocksdb/ha_rocksdb.cc (hunk 4 above)
Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc]: Really? This is weird, the test succeeds for me in a debug build. I assumed that the ON/OFF/YES/NO bool transformation was already done before coming into the check. Debugging a bit into it and yeah, you are correct, it is coming in to the check function as a string type when ON/OFF/YES/NO and not an integer type, the integer transformation just converts the string to '0' and does not recognize the ON/YES as being worth '1'. This is a good catch and in fact makes me think we have a bug over in TokuDB where it is using a similar technique to 'fail' a SET. I'll build up a fix and update when I have re-pushed.

Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc]: OK, changes made, re-tested and pushed. Thanks!

**Review by hermanlee — APPROVED** [review]
Thanks!
